### ISSUE TITLE
gccgo: offical support for musl libc

### ISSUE THREAD

**Issue by nmeum** [golang/go#51280]
**tl;dr** At [Alpine Linux](https://alpinelinux.org) downstream, we have several patches to add support for [musl libc](https://musl.libc.org) to gccgo and gofrontend, I would like to see these patches integrated upstream. This is intended as a tracking issue to help with upstream integration.

I hope this is the right place to discuss this kind of issue. If gccgo-related stuff is better discussed on the [gofrontend-dev ML](https://groups.google.com/g/gofrontend-dev) please let me know. I am just opening this here because [related gccgo issues](https://github.com/golang/go/issues/30538) were also discussed in this bug tracker.
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.5 gccgo (Alpine 11.2.1_git20220117) 11.2.1 20220117 linux/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/soeren/.cache/go-build"
GOENV="/home/soeren/.config/go/env"
GOEXE=""
GOFLAGS="-buildmode=pie"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/soeren/src/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/soeren/src/go"
GOPRIVATE=""
GOPROXY="direct"
GOROOT="/usr"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/libexec/gcc/x86_64-alpine-linux-musl/11.2.1"
GOVCS=""
GOVERSION="go1.16.5 gccgo (Alpine 11.2.1_git20220117) 11.2.1 20220117"
GCCGO="/usr/bin/gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1521489615=/tmp/go-build -gno-record-gcc-switches -funwind-tables"
</pre></details>

### What did you do?

I am a contributor to the [Alpine Linux](https://alpinelinux.org) distribution. I recently became interested in bootstrapping our Google Go implementation using gccgo. While attempting to do so, I ran into several issues since musl doesn't seem to be supported by gccgo at the moment. To resolve these issues, several patches were written for gcc and the gofrontend. These patches are currently applied at Alpine Linux downstream. Ideally, I would like to see them integrated upstream to have official support for musl libc in gccgo. Some of the patches definitely require a cleanup, however, they are presently sufficient to [bootstrap Go using gccgo on Alpine](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/27944) on all architectures supported by Alpine (aarch64, armhf, armv7, ppc64le, s390x, x86 and x86_64). The full list of patches we presently apply for gcc can be found here: https://gitlab.alpinelinux.org/alpine/aports/-/tree/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc patches specific to gofrontend/gccgo will be further discussed below.

#### Signal 34 is special on musl

musl uses signal 34 internally (similar to how glibc uses signal 32 and signal 33). For this reason, special handling is needed for this signal in the runtime. The Google Go implementation already handles signal 34 accordingly (see #39343 etc) but unfortunately gofrontend doesn't. At Alpine, we apply two patches to adjust the gofrontend signal handling code accordingly:

1. [`0033-gcc-go-Fix-handling-of-signal-34-on-musl.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0033-gcc-go-Fix-handling-of-signal-34-on-musl.patch)
2. [`0035-gcc-go-signal-34-is-special-on-musl-libc.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0035-gcc-go-signal-34-is-special-on-musl-libc.patch)

These two patches could IMHO be applied as is.

#### Definition of `off64_t` / `loff_t` as a macro

musl defines the non-standard `off64_t` and `loff_t` types as a macro, not as `typedef`. The gofrontend code uses these types unconditionally in a [few](https://github.com/golang/gofrontend/blob/4bdff733a0c2a9ddc3eff104b1be03df058a79c4/libgo/go/syscall/libcall_linux.go#L214-L235) [places](https://github.com/golang/gofrontend/blob/4bdff733a0c2a9ddc3eff104b1be03df058a79c4/libgo/mksysinfo.sh#L406-L410). Unfortunately, `-fdump-go-spec` (and `mksysinfo.sh`) are not able to correctly detect the presence of these types when they are defined as macros as not as `typedef`. For this reason, they are absent from the Go type declaration file generated by `-fdump-go-spec` and thus result in libgo compilation errors.

At Alpine Linux downstream, we presently fix this via the [`0041-libgo-Recognize-off64_t-and-loff_t-definitions-of-mu.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0041-libgo-Recognize-off64_t-and-loff_t-definitions-of-mu.patch) by `#undef`'ing the macros and defining 
`off64_t` / `loff_t` manually using `typedef`. Alternatively, the libgo code could probably be rewritten to not rely on the presence of these types since they are IIRC non-standard (probably the cleaner solution).

#### XSI-compliant `strerror_r()`

When libgo is compiled on Linux, it is assumed that the GNU-extended version of `strerror_r(3)` is available for the `Errstr()` implementation. However, musl only supports the standard XSI-compliant version of `strerror_r(3)`. At Alpine Linux, we fix this via the [`0038-Use-generic-errstr.go-implementation-on-musl.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0038-Use-generic-errstr.go-implementation-on-musl.patch) by unconditionally using the XSI-compliant version on Linux. If continued support for the GNU-extended version of `strerror_r(3)` is desired it would probably be required to somehow detect musl in the Go `//+ build` flags.

#### Memory allocation issues on 32-bit arches

On 32-bit architectures gccgo's runtime has some weird memory allocation issues where it fails to allocate any memory at all with the following error message on musl at runtime:

```
fatal error: runtime: cannot allocate memory
runtime stack:
runtime.dopanic__m
	:0
runtime.throw
	:0
	:0
	:0
runtime.systemstack
	:0
runtime.addrRanges.init
	:0
runtime.pageAlloc.init
	:0
runtime.mheap.init
	:0
runtime.schedinit
	:0
	:0
	:0
```

I am presently not entirely sure what is causing this, but changing the `sysMmap` and `mmap` offset function argument type from `uintptr` to `int64` helps for some reason (see [`0044-gcc-go-Use-int64-type-as-offset-argument-for-mmap.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0044-gcc-go-Use-int64-type-as-offset-argument-for-mmap.patch)). This needs further investigation, any idea what might be causing this? On our 32-bit builders, we are also running into some out-of-memory issues when bootstrapping Go with gcc-go without setting `GOMAXPROCS=1` which may or may not be related to this issue.

#### libucontext related issues

musl itself does not support `swapcontext`, `makecontext`, etc. These are instead provided by [libucontext](https://github.com/kaniini/libucontext) which requires a few minimal gccgo patches to link against libucontext and fix some minor quirks in this regard:

1. [0037-gcc-go-link-to-libucontext.patch](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0037-gcc-go-link-to-libucontext.patch)
2. [0036-gcc-go-undef-SETCONTEXT_CLOBBERS_TLS-in-proc.c.patch](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0036-gcc-go-undef-SETCONTEXT_CLOBBERS_TLS-in-proc.c.patch)
3. [0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch) (this [has been submitted upstream already](https://gcc.gnu.org/pipermail/gcc-patches/2022-January/587520.html))

Maybe it is possible to detect use of libucontext via GNU autotools and adjust the buildsystem configuration accordingly?

#### gcc issues

We also ran into two issues which are more closely related to gcc itself rather than gofrontend. These have already been [fixed](https://gcc.gnu.org/pipermail/gcc-patches/2021-November/585536.html) [upstream](https://gcc.gnu.org/pipermail/gcc-patches/2021-December/587138.html). The remaining patches, which have not been integrated, are more closely related to gofrontend. According to my understanding of the [gofrontend contribution guidelines](https://go.dev/doc/gccgo_contribute) I am hence opening this tracking issue here and not in the gcc bugtracker.

Would you be interested in including the discussed patches in gofrontend?

### What did you expect to see?

Upstream gccgo musl support.

### What did you see instead?

No support for musl libc in gccgo.

---

CC: @awilfox, @kaniini, @ianlancetaylor

**Comment by awilfox** [comment]
Hi there, I'm the original author of the patchset and I am happy to see this working its way upstream.  Note that while it doesn't apply cleanly to GCC 11 any more (we ship GCC 8 in Adélie right now), these changes are needed to `runtime/go-signal.c` to work on PowerPC:

* [333-gccgo-signal-ppc64.patch](https://cgit.adelielinux.org/packages/tree/system/gcc/333-gccgo-signal-ppc64.patch?h=1.0-RC2), which fixes compilation on musl, and
* [335-gccgo-signal-ppc32.patch](https://cgit.adelielinux.org/packages/tree/system/gcc/335-gccgo-signal-ppc32.patch), which expands support to 32-bit PowerPC

Also, I don't see discussion of the `struct stat` fix, which is definitely required for 32-bit systems.  That's [340-gccgo-time64-stat.patch](https://cgit.adelielinux.org/packages/tree/system/gcc/340-gccgo-time64-stat.patch) in our patchset.

Happy to do any forward-porting or testing required, including with GCC 11 or trunk if desired.

**Comment by nmeum** [comment]
>  Note that while it doesn't apply cleanly to GCC 11 any more (we ship GCC 8 in Adélie right now) […]

We have revised and updated many of the patches since, we now use them successfully with GCC 11 at Alpine (see my `go version` output above).

> […] these changes are needed to runtime/go-signal.c to work on PowerPC

The two patches you are mentioning for PPC are superseded by [`0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch).  musl nowadays defines pt_regs as an incomplete type which can be completed by inclusion of `asm/ptrace.h`.

> Also, I don't see discussion of the `struct stat` fix.

Yeah, I missed that one. Will update the discussion accordingly.

**Comment by ianlancetaylor** [comment]
Thanks.  Discussing here is fine.  Ideally send patches to the gofrontend repo as described at https://go.dev/doc/contribute.

For `mmap` the right fix may be to use `_off_t` in the `//extern` line.  The `mmap` function itself should keep the same signature.

Detecting `libucontext` via autotools won't work when cross-compiling.

**Comment by awilfox** [comment]
> We have revised and updated many of the patches since

Right, but I didn't see the PPC regs patch - that patch is what no longer applies cleanly.

> The two patches you are mentioning for PPC are superseded by [`0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch`](https://gitlab.alpinelinux.org/alpine/aports/-/blob/da8ed6612f864a639e8f381ebae7dab7c20164a2/main/gcc/0046-libgo-include-asm-ptrace.h-for-pt_regs-definition-on.patch).

Ah, I missed that.

**Comment by awilfox** [comment]
> Detecting `libucontext` via autotools won't work when cross-compiling.

I think it would.  These conditions would be true on musl:

* We are on `linux-musl`.
* Usage of `getcontext` fails to link.  This check handles the case that if musl adds support later, libucontext would no longer be necessary.

If those are true, we need to add `-lucontext`.  Presence detection probably isn't necessary, because if it was missing, the error would be obvious.

**Comment by nmeum** [comment]
> Ideally send patches to the gofrontend repo as described at https://go.dev/doc/contribute.

I think this will be difficult in this case since the Alpine patchset is authored by multiple people, I can't sign the CLA for everyone (and also I don't have a Google account myself). @awilfox would you be interested in submitting your patches yourself as a starting point?

Alternatively, maybe @ianlancetaylor could use our patches as a source of inspiration and commit modified version of them, i.e. treat this more as a bug report (as I mentioned most need some cleanup anyhow)? I also think that many of these 2-3 line changes are copyrightable in the first place.

**Comment by ianlancetaylor** [comment]
When cross-compiling we should not assume that links succeed in general.  People need to be able to build GCC in order to build libc.  So when cross-compiling a library we should be prepared to assume that all link tests will fail.

For patches that are just a couple of lines I agree that they aren't copyrightable and it doesn't matter who sends them into Gerrit.  Sending to gcc-patches is also OK but less convenient; be sure to CC me for any patches to libgo or gcc/go, as otherwise it's easy for me to miss them.

It's unrealistic to wait for me to write my own version of the patches, I don't use musl and I don't have the time.  Sorry.

**Comment by kaniini** [comment]
I'm willing to handle this, I can submit these patches on behalf of Alpine via Chainguard, who have a CLA on file with Google :)

**Comment by awilfox** [comment]
Assuming proper attribution for the patches I authored, I am completely fine with @kaniini handling the submissions of these patches.  I have already signed the Google CLA individually for google/marl#7 which should still apply for the patches I wrote for GCC Go.

**Comment by nmeum** [comment]
Note to myself: With musl 1.2.3 https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105225 is also needed

**Comment by nmeum** [comment]
I am happy to report that the majority of the Alpine gccgo musl patches has been integrated upstream. Apart from the libucontext issues discussed above, the only remaining portability issue I am presently aware of is that `libgo/go/syscall/errstr_glibc.go` is used by default on Linux and [unconditionally uses the `strerror_r()` GNU-version](
https://github.com/golang/gofrontend/blob/bbb3a4347714faee620dc205674510a0f20b81ae/libgo/go/syscall/errstr_glibc.go#L11) which musl does not provide, musl only provides the XSI-compliant version. For this reason, we have an Alpine downstream patch which forces use of `libgo/go/syscall/errstr.go` over `libgo/go/syscall/errstr_glibc.go` [\[1\]](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/main/gcc/0034-Use-generic-errstr.go-implementation-on-musl.patch).

Any idea how this could be fixed in a portable way upstream? GNU autoconf provides [`AC_FUNC_STRERROR_R`](https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Particular-Functions.html#index-AC_005fFUNC_005fSTRERROR_005fR-516) to detect if `strerror_r()` returns a `char*` or an `int`, but using that would probably require moving the `strerror()` post-processing from Go to C code.

**Comment by ianlancetaylor** [comment]
Moving from Go to C for something like this is fine.  That is, we can fix this by testing for the functionality in libgo/configure.ac and writing the relevant code in C using `#ifdef` to call the right function.  For example, we already define `strerror_r` in runtime/go-nosys.c.

**Comment by nmeum** [comment]
The `strerror_r` compatibility issue [has been fixed as well](https://github.com/golang/gofrontend/commit/fef6aa3c1678cdbe7dca454b2cebb369d8ba81bf). The only patch that remains at Alpine downstream is the `libucontext` patch. Would there be an interest in resolving this too? Maybe this could be done through an additional `./configure` flag like `--enable-libucontext` or something? Alternatively, maybe something along the lines of ` AC_SEARCH_LIBS([swapcontext], [c ucontext], [], AC_MSG_ERROR([swapcontext required but not found]))` could work too?

Also: Is there a CI for gofrontend/gccgo? Would there be an interested in adding Alpine to it?

**Comment by ianlancetaylor** [comment]
Using `AC_SEARCH_LIBS` sounds like the right approach.  It could be something like first test for `makecontext` and then if it is not found search for `-lucontext`.  Note that `swapcontext` is not used, only `makecontext`, `getcontext`, and `setcontext`.

Note that on x86_64 we use custom written assembly code for `makecontext`, `getcontext`, and `setcontext`, defined in runtime/go-context.S.  In an ideal world we would use custom assembly code for all targets.

Alas, there is no CI for gofrontend.

**Comment by ianlancetaylor** [comment]
Thanks for all the work on this.
